keps/993-two-phase-admission/README.md

```diff
@@ -0,0 +1,246 @@
+	RetryDelayMinutes *int64. `json:"retryDelayMinutes,omitempty"`
+
+	// A reference to the additional parameters for the check. 
+	Parameters *AdmissionCheckParametersReference `json:"parameters,omitempty"`
+
+	// If set to true, the workload will preempt other and block
+	// capacity before the check is passed.
+	BookCapacity bool `json:"bookCapacity,omitempty"`
+}
+
+// Points to where the parameters for the checks are. 
+// As clusterqueue are in cluster scope - this should be
+// a dedicated CRD specific for the Controller.
+type AdmissionCheckParametersReference struct {
+	// ApiGroup is the group for the resource being referenced. 
+	APIGroup string `json:"apiGroup"`
+
+	// Kind is the type of the resource being referenced.
+	Kind string `json:"kind""`
+
+	// Name is the name of the resource being referenced.
+	Name string `json:"name"`
+}
+```
+
+For every workload that is put to a ClusterQueue that has admissionChecks
+configured Kueue will add admissionCheckStatuses:
+
+* “AdmissionPrecheck” to false
+* “<checkName>” to unknown for each of the AdmissionChecks
+
+Kueue will perform the very same checks that it does today, 
+before admitting a workload. However, oncethe basic checks 
+pass AND there are some AdmissionChecks configured, AND the
+workload is not in on-hold retry state from some check, it will:
+
+1. Fill the Admission field in workload, with the desired flavor assignment. 
+2. Not do any preemptions yet (unless BookCapacity is set to true).
+3. Set “AdmissionPrecheck” admisisonChecksStatuses  to true.
+
+Kueue will only pass as many pods past AdmissionPrecheck as there would
+fit in the quota together, possibly limited by a Kueue configuration option.
+That would violate a bit BestEffort logic of queues - if there is 1000 tasks
+in a queue that doesn’t pass quota and 1001st task passes, the best effort
+queue would let it in. BestEffort queue with 1000 tasks that don’t pass
+AdmissionChecks would not let 1001st task in (it will not reach the
+checks). Without this limitation, a large number of tasks could be switched
+back and forth from Precheck to suspended state. 
+
+
+Then it will wait until all additional check conditions are satisfied (set
+to true). Once it happens, the Kueue will recheck that Admission settings
+are still valid (check quota/preemptions/reclamation) and admit the
+workload (doing preemptions if needed). Kueue will also recheck all of the
+prechecked workloads on every other workload admission, to make sure
+that their flavor admission is still possible. 
+
+If any of these checks fails (either on the final check or when admission
+other workloads), the workload gets back to the suspended state, and gets
+their AdmissionChecks and AdmissionPrecheck cleared.
+
+If any check is switched to Reject or Retry, the workload is either
+completely rejected or switched back to suspend state (with exponential
+backoff retry delay). QuotaExhausted means that the quota is reached
```

Comment:
this might be confusing, as quota is a term in Kueue. IIUC, here you are referring to an autoscaling quota.

Maybe we can be more generic and say: `BookQuotaRequired`, similar to the `BookQuota` field in the AdmissionCheckSpec

Reply:
Ok, changed.